Fixes ability to reference data on disk multiple times

johnedquinn opened the pull request:
## Relevant Issues
- Closes #936 
- Closes #518 

## Issue Description
- Whenever we pass in an input file, the data cannot be access twice due to how we use sequences.
  - When we pass in a file, we essentially use the IonReader to create a sequence of IonValues
  - Then, we create a `newBag` with the resulting sequence.
- The above flow would not allow referencing data twice because it is a sequence (can only be consumed once)

## Included Changes
- This PR proposes that we establish a re-openable stream and an implementation of Iterable that uses the re-openable stream.
  - So, in the context of this PR, we take in a file, create a re-openable stream (InputSource), wrap the stream with the new iterable (InputSourceIterable for Ion files, DelimitedFileIterable for CSV files), and use it for the creation of the ExprValue.
  - With this, when we want to iterate a file's values multiple times, we are able to start at the beginning of the file each time.

## Other Changes
- Since we've revamped how we're handling streams, this implementation also resolves our inability to close file streams seen in #518
- I've also added testing to test joins between Ion files containing single values and sequences of values. See `CliTest.kt`

## Example Usage
- See the updated usage of `read_file`:
<img width="906" alt="Screen Shot 2023-01-09 at 10 47 18 AM" src="https://user-images.githubusercontent.com/40360967/211384472-e15a600c-d193-49bb-ab10-5d9e2c7e9c4b.png">

- See the updated usage of using `--input`:
<img width="1121" alt="Screen Shot 2023-01-09 at 10 49 59 AM" src="https://user-images.githubusercontent.com/40360967/211384916-82369f55-02ef-41e8-a423-ae766c1639a9.png">

- As you can see, there aren't any issues regarding consumption of the input sequence.

## Other Information
- Updated Unreleased Section in CHANGELOG: **YES**
- Any backward-incompatible changes? **NO**
- Any new external dependencies? **NO**

## License Information

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

vgapeyev commented on partiql-app/partiql-cli/src/main/kotlin/org/partiql/cli/InputSource.kt:
```diff
@@ -0,0 +1,31 @@
+    internal class InputStreamSource(private val stream: InputStream) : InputSource() {
+        override fun stream(): InputStream = stream
+    }
+}
```
The way I got to understand `InputSource`, it is a stream that can be re-opened multiple times while (like any stream) hiding away the source of data (a file or whatever).  I think this can become a pretty neat abstraction, so here are a few thoughts on how it could be improved.  

First, it does not quite fulfill the above idea: while `FileStreamSource` is "re-openable" (at each call to the `stream()` method, `InputStreamSource` is not.  However, `InputStreamSource` is currently instantiated only once (at CliTestUtility:30), from a string.  So, I think it can be re-done to follow `FileStreamSource`, except instantiating from a string rather than a file -- therefore becoming re-openable as well.  
Then better names for both would be `FileSource` and `StringSource`, I'd suggest. 

Second, the abstraction is "leaky" currently.  Ideally, one would create an `InputSource` (from a file or a string, or whatever) and then only use methods like `stream()` that do not reveal anything about the underlying source.  That would be a "tight" abstraction we are striving for.  However, in `Cli.runWithIonInputWrapped` it discriminates on the underlying source and, in the file-based case even gets to the underlying file.  There is also some duplication of code there and in `IonFileConverterIterable`.  
It is my hope that it can be refactored, probably moving some of the duplicated code to somewhere here in `InputSource`.  The goal is to have in `runWithIonInputWrapped` a simple call like `input.stream()` -- with no discrimination on the underlying input there, in `runWithIonInputWrapped`.
I admit I did not think through all of the details, so let me know if there is an obstacle that I overlooked of if it is not clear what I am suggesting.

vgapeyev reviewed (commented):
Here are a few minor suggestions and a bigger ask: could it be possible to add a test or two with the examples currently in the PR description? That is, tests for the situation of self-joins in #936 that prompted this work.  At first I thought that the need to read data from files was an obstacle here, but I see a few tests there that just create the needed files before reading them in, so this should be doable.

johnedquinn commented on lang/src/test/resources/expr-value-factory-tests/bag.ion:
```diff
@@ -0,0 +1,4 @@
+// This is used for the ExprValueFactoryTest to test whether we can create a bag from a sequence of Ion values
+0
+1
+2
```
Oh, nice catch. That's seemingly a leftover from my first revision.

johnedquinn commented on partiql-app/partiql-cli/src/main/kotlin/org/partiql/cli/functions/ReadFile.kt:
```diff
@@ -73,18 +72,32 @@ internal class ReadFile(private val ion: IonSystem) : ExprFunction {
+                val iterable = object : Iterable<ExprValue> {
+                    override fun iterator(): Iterator<ExprValue> {
+                        return inputIonValue.iterator()
+                    }
```
Oof yes, that's just me being dumb. Thanks for that.

johnedquinn commented:
ACK. So, I went back and:
- rebased from main
- added more testing
- changed some logic to include the fixes for #518 (cleaning up file streams since we now have a better streaming flow)
- updated the PR description

LMK if that works!

vgapeyev commented on partiql-app/partiql-cli/src/main/kotlin/org/partiql/cli/utils/DelimitedFileIterable.kt:
```diff
@@ -0,0 +1,35 @@
+    override fun close() {
+        readers.forEach { it.close() }
+    }
+}
```
I am not sure what is going on here.  If I am confused, please help me understand.

The point of `DelimitedValues.exprValue` is to create a lazy sequence, while keeping the `reader` open.  In other words, even though the sequence is "created", it is not traversed until someone actually traverses it.  Until that traversal happens, the `reader` must be kept open.  
Now, the only call site of this iterator method is this, in ReadFile:
```
        val iterable = DelimitedFileIterable(ion, inputSource, csvFormatWithOptions, encoding, conversion)
        iterable.use { ExprValue.newBag(it) }
```
and it does not do the traversal, from what I understand: 
The call to `ExprValue.newBag` receives `it`, an instance of `DelimitedFileIterable` from the previous line and calls its `iterator` method from Line 26 above, which opens an iterator on the reader.  The only thing that is done after this by the code block passed to `use` is wrapping the iterator as an `ExprValue` (`BagExprValue` specifically).  Importantly, nothing is yet read from the reader.  But immediately after that, the 'use' method initiates closing the reader, through the method `close` on Line 32.  

Separately, why is the complication with keeping the list of readers?  In the above (and only) call site, the `iterator` method seems to be called only once and, therefore, only one reader is created. So, whatever the complication is for, is it being exercised?  But if it were, wouldn't `close()` close all the readers at once, regardless of whether they all have been already consumed?

vgapeyev reviewed (commented):
I got to the the point in DelimetedFileIterable and thought it is worth clarifying what's happening there prior to going further with the review.
